src/Workspaces/Core/Portable/EmbeddedLanguages/VirtualChars/VirtualCharSequence.cs

```diff
@@ -0,0 +1,147 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars
+{
+    /// <summary>
+    /// Represents the individual characters that raw string token represents (i.e. with escapes collapsed).  
+    /// The difference between this and the result from token.ValueText is that for each collapsed character
+    /// returned the original span of text in the original token can be found.  i.e. if you had the
+    /// following in C#:
+    ///
+    /// "G\u006fo"
+    ///
+    /// Then you'd get back:
+    ///
+    /// 'G' -> [0, 1) 'o' -> [1, 7) 'o' -> [7, 1)
+    ///
+    /// This allows for embedded language processing that can refer back to the user's original code
+    /// instead of the escaped value we're processing.
+    /// </summary>
+    internal partial struct VirtualCharSequence
+    {
+        public static readonly VirtualCharSequence Empty = Create(ImmutableArray<VirtualChar>.Empty);
+
+        public static VirtualCharSequence Create(ImmutableArray<VirtualChar> virtualChars)
+            => new VirtualCharSequence(new ImmutableArrayChunk(virtualChars));
+
+        public static VirtualCharSequence Create(int firstVirtualCharPosition, string underlyingData, TextSpan underlyingDataSpan)
+            => new VirtualCharSequence(new StringChunk(firstVirtualCharPosition, underlyingData, underlyingDataSpan));
+
+        /// <summary>
+        /// The actual characters that this <see cref="VirtualCharSequence"/> is a portion of.
+        /// </summary>
+        private readonly Chunk _leafCharacters;
+
+        /// <summary>
+        /// The portion of <see cref="_leafCharacters"/> that is being exposed.
+        /// </summary>
+        private readonly TextSpan _span;
```

Comment:
>_span [](start = 34, length = 5)

It would be good to clarify the meaning of `_span.End`. The comment above and the logic in `AssertAdjacentTo` suggests that `End` is not included. But the logic in the `VirtualCharSequence` constructor suggests that it is.